pkg/csi/cinder/controllerserver.go

```diff
@@ -777,12 +907,19 @@ func (cs *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.Co
 		return nil, status.Error(codes.InvalidArgument, "Volume ID not provided")
 	}
 
-	volume, err := cs.Cloud.GetVolume(volumeID)
-	if err != nil {
-		if cpoerrors.IsNotFound(err) {
-			return nil, status.Errorf(codes.NotFound, "Volume %s not found", volumeID)
+	var volume *volumes.Volume
+	var err error
+	for _, cloud := range cs.Clouds {
+		volume, err = cloud.GetVolume(volumeID)
+		if err != nil {
+			if cpoerrors.IsNotFound(err) {
+				continue
+			}
+			return nil, status.Errorf(codes.Internal, "ControllerGetVolume failed with error %v", err)
 		}
-		return nil, status.Errorf(codes.Internal, "ControllerGetVolume failed with error %v", err)
```

Comment:
So we don't know which cloud each volume is on and we need to iterate over all of them? This feels a bit inefficient here.

Reply:
Indeed !
In this case we wanna implement [ControllerGetVolumeRequest](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllergetvolume) RPC message unfortunately this message don't provide `secrets` field, so I can't use my trick describe above with [StorageClasse Secrets](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html).

And as you can figure it doesn't handle secrets for operation `ControllerGetVolume`
![image](https://github.com/kubernetes/cloud-provider-openstack/assets/4145084/f38d2cf9-864b-497f-a299-8292b6c0fb2a)

Finally I had to choose one of my 2 first ideas and I choose to iterate over each cloud.
I don't know OS call complexity but if it is a hash table or equivalent this call should be o(1) so iterate over cloud provide a linear complexity o(n) with n the number of cloud configured.
I find this acceptable, the best way should probably to add support of operation `ControllerGetVolume` in [external-provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html) to being able to propagate cloud name in RPC call and retrieve a complexity o(1) (or flat OS complexity). But I think that is a hard work for a low gain.